pallets/parachain-staking/src/migrations.rs

```diff
@@ -17,142 +17,120 @@
 // If you feel like getting in touch with us, you can do so at info@botlabs.org
 
 use frame_support::{
-	traits::{
-		fungible::freeze::Mutate as MutateFreeze, Get, GetStorageVersion, LockIdentifier, LockableCurrency,
-		OnRuntimeUpgrade, ReservableCurrency, StorageVersion,
-	},
-	weights::Weight,
+	pallet_prelude::DispatchResult,
+	traits::{fungible::freeze::Mutate as MutateFreeze, LockIdentifier, LockableCurrency},
 };
-use pallet_balances::{BalanceLock, Freezes, IdAmount, Locks};
-use sp_runtime::SaturatedConversion;
-use sp_std::marker::PhantomData;
+use pallet_balances::{BalanceLock, Freezes, Locks};
+use sp_runtime::{SaturatedConversion, WeakBoundedVec};
 
 use crate::{
 	types::{AccountIdOf, CurrencyOf},
-	Config, FreezeReason, Pallet, STORAGE_VERSION as TARGET_STORAGE_VERSION,
+	Config, FreezeReason,
 };
 
-const STAKING_ID: LockIdentifier = *b"kiltpstk";
+pub const STAKING_ID: LockIdentifier = *b"kiltpstk";
 
-const CURRENT_STORAGE_VERSION: StorageVersion = StorageVersion::new(8);
-
-pub struct BalanceMigration<T>(PhantomData<T>);
-
-impl<T: Config> OnRuntimeUpgrade for BalanceMigration<T>
+pub fn update_or_create_freeze<T: Config>(user_id: &T::AccountId) -> DispatchResult
 where
-	<T as Config>::Currency: ReservableCurrency<T::AccountId>,
-	<T as Config>::Currency: LockableCurrency<T::AccountId>,
+	CurrencyOf<T>: LockableCurrency<AccountIdOf<T>>,
 {
-	fn on_runtime_upgrade() -> frame_support::weights::Weight {
-		log::info!("Staking: Initiating migration");
-
-		let onchain_storage_version = Pallet::<T>::on_chain_storage_version();
-		if onchain_storage_version == CURRENT_STORAGE_VERSION {
-			TARGET_STORAGE_VERSION.put::<Pallet<T>>();
-			<T as frame_system::Config>::DbWeight::get()
-				.reads_writes(1, 1)
-				.saturating_add(do_migration::<T>())
-		} else {
-			log::info!(
-				"Staking: No migration needed. This file should be deleted. Current storage version: {:?}, Required Version for update: {:?}", 
-				onchain_storage_version,
-				CURRENT_STORAGE_VERSION
-			);
-			<T as frame_system::Config>::DbWeight::get().reads(1)
-		}
-	}
-
-	#[cfg(feature = "try-runtime")]
-	fn pre_upgrade() -> Result<sp_std::vec::Vec<u8>, &'static str> {
-		use sp_runtime::traits::Zero;
-		use sp_std::vec;
-
-		let count_freezes = pallet_balances::Freezes::<T>::iter().count();
-		assert!(count_freezes.is_zero(), "Staking Pre: There are already freezes.");
-
-		assert_eq!(Pallet::<T>::on_chain_storage_version(), CURRENT_STORAGE_VERSION);
-
-		log::info!("Staking: Pre migration checks successful");
-
-		Ok(vec![])
-	}
-
-	#[cfg(feature = "try-runtime")]
-	fn post_upgrade(_pre_state: sp_std::vec::Vec<u8>) -> Result<(), &'static str> {
-		use sp_runtime::traits::Zero;
-
-		let count_freezes = pallet_balances::Freezes::<T>::iter().count();
-
-		assert!(!count_freezes.is_zero(), "Staking: There are still no freezes.");
-
-		assert_eq!(Pallet::<T>::on_chain_storage_version(), TARGET_STORAGE_VERSION);
+	let locks: WeakBoundedVec<
```

Comment:
Do you need to specify the full type here?

Reply:
removed:  [9f04aa3](https://github.com/KILTprotocol/kilt-node/pull/540/commits/9f04aa35b0967cd077370e1a47d0bedb33cf764f)